### PR TITLE
Update GPU docs about fitting in memory

### PR DESCRIPTION
**Pull request by turian**
## What does this PR do?

PL's automatic GPU behavior is not well documented. This PR explains how data is loaded into the GPU, and how to avoid doing it every batch

Fixes # (issue)

https://github.com/PyTorchLightning/pytorch-lightning/issues/2361

# Before submitting

- [n/a ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [Y] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [Y] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you create a separate PR for every change.
- [Y] Did you make sure to update the documentation with your changes?
- [n/a] Did you write any new necessary tests? 
- [n/a] Did you verify new and existing tests pass locally with your changes?
- [n/a] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

### CODE REVIEW

File: docs/source/single_gpu.rst [hunk 1]
```diff
@@ -11,4 +11,11 @@ there's no need to set them yourself.
     :skipif: torch.cuda.device_count() < 1
 
     # train on 1 GPU (using dp mode)
-    trainer = Trainer(gpus=1)
\ No newline at end of file
+    trainer = Trainer(gpus=1)
+
+Note that each step will load the batch from the CPU to the GPU.
```
Comment (williamFalcon) [inline docs/source/single_gpu.rst:16]: this is standard PyTorch practice, I'm not sure it needs to be spelled out

**Comment by williamFalcon** [comment]
Thanks for the PR!

I don't think this comment is needed.

1. CPU to GPU transfer is standard PyTorch practice.
2. This is a very particular case where the data fits in GPU memory.
3. Lightning is device agnostic... this means that calling .cuda() directly is discouraged since it won't work for TPUs or if you want to keep on CPU only.

However, i like the general addition of helpful comments for future things as you get more familiar with the framework!

I'm going to recommend that this not be merged.

**Comment by awaelchli** [comment]
I agree with William, this does not affect 99% of users and by changing one single trainer arg you can train on TPU, multi-gpu etc which will break your dataloading if it is device specific. Besides, if you used memory pinning and non blocking transfers, would it even make a difference?

**Comment by turian** [comment]
Thank you, I understand. Just a few more questions, for my edification:

*Lightning is device agnostic... this means that calling .cuda() directly is discouraged since it won't work for TPUs or if you want to keep on CPU only.*

@williamFalcon what is the best practice then in lightning, if you want to store all your data in the CUDA memory before training?

*Besides, if you used memory pinning and non blocking transfers instead, would it even make a difference?*

@awaelchli thank you for the pointers. But pinning memory and non-blocking transfers still will involve CPU to GPU transfers, correct? If for example, I'm in a setting where the nets are not too big, data transfer could still slow things down?

**Comment by maxwass** [comment]
@turian did you figure out a nice way to do this?

I have a use case where I would like all my data to reside permanently on GPU and not waste time with the CPU->GPU transfers. @williamFalcon @awaelchli Is there a recommended way to do this, even if we don't want to codify it in the documentation?

**Comment by turian** [comment]
@maxwass I am not completely sure I have this solved correctly. I moved everything to the GPU in my code but there is still very low GPU usage for some of my models.

I would like a simple explanation for how to keep EVERYTHING on the GPU in lightning, without tears or fear. @williamFalcon @awaelchli

**Comment by maxwass** [comment]
I cry to thee, oh kind and merciful PL overlords @williamFalcon @awaelchli, help us avoid such tears and fears. 

Other than a standard way to keep data on GPU, I would think other things may need to be considered as well. In my case, and likely in others, the epochs will be super fast, so for good performance I will likely need to limit any other sort of per_step/epoch CPU<->GPU communications, i.e. logging (log_every_n_steps, flush_logs_every_n_steps, etc) and other nitty gritty operations I may not be aware of.

I of course just do straight pytorch, but I would lose all the great features/tools in this environment.

**Comment by turian** [comment]
Agreed. There is some per-epoch slowness that is hard to toggle without know ptl internals. Even then I am not sure I cannot push the optimizations further and get closer to pure torch's speed. But I prefer the ptl environment too.

**Comment by awaelchli** [comment]
Hello

Could we bring the discussion to a GitHub [discussion](https://github.com/PyTorchLightning/pytorch-lightning/discussions). It will be more discoverable compared to this very old PR :) 

> I have a use case where I would like all my data to reside permanently on GPU and not waste time with the CPU->GPU transfers. @williamFalcon @awaelchli Is there a recommended way to do this, even if we don't want to codify it in the documentation?

A bit more information would be needed here. What exactly do you mean by "permanently"? 
If you have a dataset as a full tensor, you could move it to GPU all at once, given of course that it all fits into GPU memory. This could be done before `trainer.fit()` or inside of a hook like `on_train_start()`.

@turian Thanks for creating the issue. Lightning should not add any noticeable overhead, otherwise it's a bug or perhaps due to abnormal usage.

**Comment by turian** [comment]
@awaelchli yes, see the discussion I created: https://github.com/PyTorchLightning/pytorch-lightning/issues/10660

**Pull request closed without merging**